### PR TITLE
refactor(stats): modernize `statsd::StatsdClient`, make global `unique_ptr`

### PR DESCRIPTION
**Pull request by kwvg**
## Additional Information

Support for transmitting stats to a Statsd server has been courtesy of Statoshi ([repo](https://github.com/jlopp/statoshi)), implemented Dec, 2020 by [dash#2515](https://github.com/dashpay/dash/pull/2515) but since then, it hasn't gotten much attention aside from benefiting from codebase-wide changes and the occasional compiler appeasement. This pull request aims to give our statistics code some TLC.

Changes include:

* Limiting initialization to solely during construction and moving the responsibility of fetching arguments outside of `statsd::StatsdClient`.
* Using the RAII `Socks` wrapper as early as possible (we still need to construct a raw socket ourselves but this is done in the initializer and control is moved to the wrapper and everywhere else, the wrapper is used) 
* Utilizing existing networking code to generate the socket address
  * This lets us trivially allow IPv6 connections as the responsibility to construct it safely is moved to `CService`.
* Using `std::string` and our string manipulation capabilities (replacing `snprintf` with `strprintf`), replacing platform-specific types (replacing `short` with `uint16_t`).

## Breaking Changes

None observed.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas **(note: N/A)**
- [x] I have added or updated relevant unit/integration/functional/e2e tests **(note: N/A)**
- [x] I have made corresponding changes to the documentation **(note: N/A)**
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

File: src/statsd_client.cpp [hunk 1]
```diff
@@ -64,16 +64,8 @@ StatsdClient::StatsdClient(const std::string& host, const std::string& nodename,
...
-    if (hSocket == INVALID_SOCKET) {
-        LogPrintf("Cannot create socket (socket() returned error %s), cannot init StatsdClient\n",
-                  NetworkErrorString(WSAGetLastError()));
-        return;
-    }
-    m_sock = std::make_unique<Sock>(hSocket);
-
     CNetAddr netaddr;
-    if (!LookupHost(m_host, netaddr, /*fAllowLookup=*/true) || !netaddr.IsIPv4()) {
+    if (!LookupHost(m_host, netaddr, /*fAllowLookup=*/false) || !(netaddr.IsIPv4() || netaddr.IsIPv6())) {
```
Comment (PastaPastaPasta) [inline src/statsd_client.cpp:68]: nit: separate error message for ipv4 / ipv6 not

File: src/statsd_client.cpp [hunk 2]
```diff
@@ -64,16 +64,8 @@ StatsdClient::StatsdClient(const std::string& host, const std::string& nodename,
         return;
     }
 
-    SOCKET hSocket = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
```
Comment (PastaPastaPasta) [inline src/statsd_client.cpp]: nit: e2951fec6dc7e3de5a8a77276d0b97b4d4435130 should be a feat

File: src/statsd_client.cpp [hunk 3]
```diff
@@ -32,217 +37,150 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
 std::unique_ptr<statsd::StatsdClient> g_stats_client;
 
 namespace statsd {
-
-inline bool fequal(float a, float b)
+bool StatsdClient::ShouldSend(float sample_rate)
 {
-    const float epsilon = 0.0001;
-    return ( fabs(a - b) < epsilon );
+    LOCK(cs);
```
Comment (PastaPastaPasta) [inline src/statsd_client.cpp:51]: nit: minimize scope

File: src/statsd_client.cpp [hunk 4]
```diff
@@ -32,217 +37,150 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
-
-    config(gArgs.GetArg("-statshost", DEFAULT_STATSD_HOST), gArgs.GetArg("-statsport", DEFAULT_STATSD_PORT), gArgs.GetArg("-statsns", DEFAULT_STATSD_NAMESPACE));
-
-    d->sock = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
-    if ( d->sock == INVALID_SOCKET ) {
-        snprintf(d->errmsg, sizeof(d->errmsg), "could not create socket, err=%m");
-        return -1;
+    CNetAddr netaddr;
+    if (!LookupHost(m_host, netaddr, /*fAllowLookup=*/false) || !(netaddr.IsIPv4() || netaddr.IsIPv6())) {
+        LogPrintf("Unable to lookup host %s, cannot init StatsdClient\n", m_host);
```
Comment (PastaPastaPasta) [inline src/statsd_client.cpp:69]: nit: add "ERROR" to text or something similar?

File: src/statsd_client.cpp (hunk 3 above)
Comment (PastaPastaPasta) [inline src/statsd_client.cpp:51]: nit: check m_sock first so we don't do a bunch of logic and then realize we aren't enabled :)

File: src/statsd_client.h [hunk 5]
```diff
@@ -1,15 +1,20 @@
...
+
 #include <string>
 #include <memory>
 
 static const bool DEFAULT_STATSD_ENABLE = false;
-static const int DEFAULT_STATSD_PORT = 8125;
+static const uint16_t DEFAULT_STATSD_PORT = 8125;
 static const std::string DEFAULT_STATSD_HOST = "127.0.0.1";
 static const std::string DEFAULT_STATSD_HOSTNAME = "";
 static const std::string DEFAULT_STATSD_NAMESPACE = "";
```
Comment (PastaPastaPasta) [inline src/statsd_client.h:20]: nit constexpr stuff

**Review by PastaPastaPasta — COMMENTED** [review]
Generally LGTM: 007e76b10bf713a7e70fb3abcc2e6c2dc10720f2

added a few nits / suggestions

File: src/statsd_client.cpp (hunk 1 above)
Comment (kwvg) [reply src/statsd_client.cpp]: Resolved in latest push!

File: src/statsd_client.cpp (hunk 2 above)
Comment (kwvg) [reply src/statsd_client.cpp]: Resolved in latest push!

File: src/statsd_client.cpp (hunk 3 above)
Comment (kwvg) [reply src/statsd_client.cpp]: Resolved in latest push!

File: src/statsd_client.cpp (hunk 4 above)
Comment (kwvg) [reply src/statsd_client.cpp]: Resolved in latest push!

File: src/statsd_client.cpp (hunk 3 above)
Comment (kwvg) [reply src/statsd_client.cpp]: Resolved in latest push!

File: src/statsd_client.h (hunk 5 above)
Comment (kwvg) [reply src/statsd_client.h]: Resolved in latest push!

File: src/init.cpp [hunk 6]
```diff
@@ -289,6 +289,7 @@ void PrepareShutdown(NodeContext& node)
     node.banman.reset();
     node.addrman.reset();
     node.netgroupman.reset();
+    ::statsClient.reset();
```
Comment (PastaPastaPasta) [inline src/init.cpp]: nit: `stats:` is not a valid scope

**Review by UdjinM6 — APPROVED** [review]
utACK cc998abec16bfb1f2d55d92e4c0e6a495576c775

**Pull request merged**